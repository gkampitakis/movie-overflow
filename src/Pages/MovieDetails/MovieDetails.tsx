import React, { useEffect, useState } from 'react';
import { RouteComponentProps, useHistory } from 'react-router-dom';
import { AiOutlineClockCircle } from 'react-icons/ai';
import Placeholder from '../../assets/placeholder.jpeg';
import { MovieDetails } from '../../types';
import { Loader } from '../../Components';
import { getMovie } from '../../Api';
import './movieDetails.scss';

const imgSrc = (path: string, width: '185' | '92') =>
  path ? `https://image.tmdb.org/t/p/w${width}${path}` : Placeholder;
const openImage = (path: string) =>
  window.open(`https://image.tmdb.org/t/p/original${path}`, '_blank');

export default function _MovieDetails(
  props: RouteComponentProps<{ id: string }>
) {
  const history = useHistory();
  const [loading, setLoading] = useState(false);
  const { id } = props.match.params;
  const [movieDetails, setMovieDetails] = useState<MovieDetails>();

  const handleRequest = () => {
    setLoading(true);

    getMovie(id)
      .then((data) => {
        setMovieDetails(data);
      })
      .catch((error) => {
        if (error.status === 404) {
          return history.push({
            pathname: '/notFound'
          });
        }
        console.error(error);
      })
      .finally(() => setLoading(false));
  };

  const goTo = (id: string) => {
    history.push({
      pathname: `/person/${id}`
    });
  };

  useEffect(() => {
    handleRequest();
  }, [id]);

  return (
    <section className="movie_details">
      {movieDetails && (
        <>
          <div className="details">
            <header>
              <h1>{movieDetails.original_title}</h1>
              <p>
                {movieDetails.genres.map((genre) => genre.name).join(', ') ||
                  'Unknown'}{' '}
                | {movieDetails.runtime} <AiOutlineClockCircle />
              </p>
            </header>
            <article className="overview">
              <img
                src={imgSrc(movieDetails.poster_path, '185')}
                alt={movieDetails.original_title}
              />
              <div>
                <p className="tagline">{movieDetails.tagline}</p>
                <p>{movieDetails.overview}</p>
              </div>
            </article>
            <article className="general">
              <div className="released">
                <h3>{movieDetails.status}</h3>
                <p>{movieDetails.release_date}</p>
              </div>
              <div className="statistics">
                <p>Vote: {movieDetails.vote_average} âœ¨</p>
              </div>
            </article>
            {!!movieDetails.images.posters.length && (
              <>
                <hr />
                <h2 className="photos_title">Photos</h2>
                <article className="photos">
                  {movieDetails.images.posters.slice(0, 7).map((image, idx) => (
                    <img
                      onClick={() => openImage(image.file_path)}
                      key={idx}
                      src={imgSrc(image.file_path, '92')}
                      alt={movieDetails.original_title}
                    />
                  ))}
                </article>
              </>
            )}
          </div>
          {(!!movieDetails.credits.cast.length ||
            !!movieDetails.credits.cast.length) && (
            <div className="credits">
              {!!movieDetails.credits.cast.length && <h2>Cast</h2>}
              {movieDetails.credits.cast.map((person) => (
                <div onClick={() => goTo(person.id)} key={person.id}>
                  <div className="avatar">
                    <img
                      loading="lazy"
                      src={
                        person.profile_path
                          ? imgSrc(person.profile_path, '185')
                          : `https://eu.ui-avatars.com/api/?size=185&name=${person.original_name}`
                      }
                      alt={person.original_name}
                    />
                  </div>
                  <h3>{person.original_name}</h3>
                  <p>as {person.character}</p>
                </div>
              ))}
              {!!movieDetails.credits.crew.length && <h2>Crew</h2>}
              {movieDetails.credits.crew.map((person) => (
                <div onClick={() => goTo(person.id)} key={person.id}>
                  <div className="avatar">
                    <img
                      loading="lazy"
                      src={
                        person.profile_path
                          ? imgSrc(person.profile_path, '185')
                          : `https://eu.ui-avatars.com/api/?size=185&name=${person.original_name}`
                      }
                      alt={person.original_name}
                    />
                  </div>
                  <h3>{person.original_name}</h3>
                  <p>{person.job}</p>
                </div>
              ))}
            </div>
          )}
        </>
      )}
      <Loader loading={loading} />
    </section>
  );
}
