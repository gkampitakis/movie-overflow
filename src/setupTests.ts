// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import fetchMock from 'jest-fetch-mock';

fetchMock.enableMocks();

export const movieDetailsResponse = {
  id: 'mockId',
  original_title: 'mockTitle',
  runtime: 100,
  genres: [{ name: 'test' }],
  poster_path: '/mock_path',
  tagline: 'mockTagLine',
  overview: 'mockOverview',
  status: 'mockStatus',
  release_date: 'mockDate',
  vote_average: 100,
  images: {
    posters: [
      {
        file_path: '/mock_image_path'
      }
    ]
  },
  credits: {
    cast: [
      {
        id: 'id1',
        profile_path: 'cast_path',
        original_name: 'mockName',
        character: 'character1'
      },
      {
        id: 'id2',
        profile_path: null,
        original_name: 'mockName2',
        character: 'character2'
      }
    ],
    crew: [
      {
        id: 'crewId1',
        profile_path: 'crew_path',
        original_name: 'crewMockName',
        job: 'job1'
      },
      {
        id: 'crewId2',
        profile_path: null,
        original_name: 'crewMockName2',
        job: 'job2'
      }
    ]
  }
};