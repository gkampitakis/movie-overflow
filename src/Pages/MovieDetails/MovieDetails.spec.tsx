import React from 'react';
import MovieDetails from '.';
import { render, act, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FetchMock } from 'jest-fetch-mock';
import { movieDetailsResponse, mockResponseOnce } from '../../setupTests';

const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush
  })
}));

const fetchMock = fetch as FetchMock;
const mockProps = {
  match: { params: { id: 'mockId' } }
} as any;

describe('MovieDetails', () => {
  beforeEach(() => {
    fetchMock.resetMocks();
    mockHistoryPush.mockClear();
    mockResponseOnce(movieDetailsResponse);
  });
  it('Should show loading', async () => {
    fetchMock.resetMocks();
    fetchMock.mockResponseOnce(
      () =>
        new Promise<string>((resolve) => {
          setTimeout(() => resolve('{}'), 1000);
        })
    );
    await act(async () => {
      const { container } = render(<MovieDetails {...mockProps} />);

      await screen.findByTestId('loader');

      expect(container).toMatchSnapshot();
    });
  });

  it('Should render consistently', async () => {
    await act(async () => {
      const { container } = render(<MovieDetails {...mockProps} />);

      await screen.findByText(/mockTitle/i);

      expect(container).toMatchSnapshot();
    });
  });

  it('Should open new tab with image', async () => {
    const openSpy = jest.spyOn(window, 'open').mockImplementation();

    await act(async () => {
      render(<MovieDetails {...mockProps} />);

      await screen.findByText(/mockTitle/i);
      const img = screen.getAllByAltText(/mockTitle/i);

      userEvent.click(img[1]);

      expect(openSpy).toHaveBeenCalledWith(
        'https://image.tmdb.org/t/p/original/mock_image_path',
        '_blank'
      );
    });
  });

  it("Should go to person's page", async () => {
    await act(async () => {
      render(<MovieDetails {...mockProps} />);

      await screen.findByText(/mockTitle/i);
      const cast = screen.getByAltText('mockName');
      const crew = screen.getByAltText('crewMockName2');

      userEvent.click(cast);
      userEvent.click(crew);

      expect(mockHistoryPush).toHaveBeenCalledWith({
        pathname: '/person/id1'
      });
      expect(mockHistoryPush).toHaveBeenCalledWith({
        pathname: '/person/crewId2'
      });
    });
  });

  it('Should redirect to page not found', async () => {
    fetchMock.resetMocks();
    mockResponseOnce('{}', 404);
    await act(async () => {
      render(<MovieDetails {...mockProps} />);

      setImmediate(() => {
        expect(mockHistoryPush).toHaveBeenCalledWith({
          pathname: '/notFound'
        });
      });
    });
  });
});
