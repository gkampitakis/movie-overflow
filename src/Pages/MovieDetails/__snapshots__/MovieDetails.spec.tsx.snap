// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails Should render consistently 1`] = `
<div>
  <section
    class="movie_details"
  >
    <div
      class="details"
    >
      <header>
        <h1>
          mockTitle
        </h1>
        <p>
          test
           |
           
          100
           
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
            />
            <path
              d="M686.7 638.6L544.1 535.5V288c0-4.4-3.6-8-8-8H488c-4.4 0-8 3.6-8 8v275.4c0 2.6 1.2 5 3.3 6.5l165.4 120.6c3.6 2.6 8.6 1.8 11.2-1.7l28.6-39c2.6-3.7 1.8-8.7-1.8-11.2z"
            />
          </svg>
        </p>
      </header>
      <article
        class="overview"
      >
        <img
          alt="mockTitle"
          src="https://image.tmdb.org/t/p/w185/mock_path"
        />
        <div>
          <p
            class="tagline"
          >
            mockTagLine
          </p>
          <p>
            mockOverview
          </p>
        </div>
      </article>
      <article
        class="general"
      >
        <div
          class="released"
        >
          <h3>
            mockStatus
          </h3>
          <p>
            mockDate
          </p>
        </div>
        <div
          class="statistics"
        >
          <p>
            Vote: 
            100
             âœ¨
          </p>
        </div>
      </article>
      <hr />
      <h2
        class="photos_title"
      >
        Photos
      </h2>
      <article
        class="photos"
      >
        <img
          alt="mockTitle"
          src="https://image.tmdb.org/t/p/w92/mock_image_path"
        />
      </article>
    </div>
    <div
      class="credits"
    >
      <h2>
        Cast
      </h2>
      <div>
        <div
          class="avatar"
        >
          <img
            alt="mockName"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w185cast_path"
          />
        </div>
        <h3>
          mockName
        </h3>
        <p>
          as 
          character1
        </p>
      </div>
      <div>
        <div
          class="avatar"
        >
          <img
            alt="mockName2"
            loading="lazy"
            src="https://eu.ui-avatars.com/api/?size=185&name=mockName2"
          />
        </div>
        <h3>
          mockName2
        </h3>
        <p>
          as 
          character2
        </p>
      </div>
      <h2>
        Crew
      </h2>
      <div>
        <div
          class="avatar"
        >
          <img
            alt="crewMockName"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w185crew_path"
          />
        </div>
        <h3>
          crewMockName
        </h3>
        <p>
          job1
        </p>
      </div>
      <div>
        <div
          class="avatar"
        >
          <img
            alt="crewMockName2"
            loading="lazy"
            src="https://eu.ui-avatars.com/api/?size=185&name=crewMockName2"
          />
        </div>
        <h3>
          crewMockName2
        </h3>
        <p>
          job2
        </p>
      </div>
    </div>
  </section>
</div>
`;

exports[`MovieDetails Should show loading 1`] = `
<div>
  <section
    class="movie_details"
  >
    <span>
      Loading...
    </span>
  </section>
</div>
`;
